scipy/_lib/_array_api_util.py

```diff
@@ -0,0 +1,65 @@
+"""Utility functions to use Python Array API compatible libraries.
+
+For the context about the Array API see:
+https://data-apis.org/array-api/latest/purpose_and_scope.html
+
+The SciPy use case of the Array API is described on the following page:
+https://data-apis.org/array-api/latest/use_cases.html#use-case-scipy
+"""
+
+import numpy as np
+
+from scipy._lib import _pep440
+
+_NUMPY_INCLUDES_ARRAY_API = _pep440.parse(np.__version__) >= _pep440.Version(
+    "1.22.0"
+)
+
+
+def _get_namespace(*arrays):
+    """
+    Returns the module that implements Array API compatible functions or NumPy.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays to get the namespace from.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x = np.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy'
+
+    >>> import numpy.array_api as xp
+    >>> x = xp.arange(6)
+    >>> _get_namespace(x).__name__
+    'numpy.array_api'
+    """
+    # `arrays` contains one or more arrays
+    # When there is no __array_namespace__,
+    # we will use assume the default (NumPy) namespace.
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, "__array_namespace__") else np
+        for x in arrays
+    }
+
+    if len(namespaces) != 1:
+        raise ValueError(
+            f"Expected a single common namespace for array inputs, \
+              but got: {[n.__name__ for n in namespaces]}"
```

Comment:
This or we call unique and only return single occurrences of namespaces. This is to mitigate cases where user would through a large number of arrays in some functions.

Reply:
> This is to mitigate cases where user would through a large number of arrays in some functions.

I was wondering the same. May be we can do something like below,

1. Obtain namespace of first array and assume that it is the answer.
2. Then from second array onwards, if namespace doesn't change then keep moving to thrid, fourth, fifth and so on.
3. Once we encounter a different namespace raise the error.

This wouldn't require usage of sets but it won't be verbose because, we will be exiting before we know the namespaces of all the arrays so we won't be able to show all the namespaces in our error message.

Reply:
`namespaces` is a set so only a single occurrence of each namespace is printed in the error message.

Reply:
Great then!